EIPS/eip-2200.md

```diff
@@ -0,0 +1,322 @@
+---
+eip: 2200
+title: Structured Definitions for Net Gas Metering
+author: Wei Tang (@sorpaas)
+discussions-to: https://github.com/sorpaas/EIPs/issues/1
+status: Draft
+type: Standards Track
+category: Core
+created: 2019-07-18
+requires: 1283, 1706, 1884
+---
+
+## Simple Summary
+
+This is an EIP that implements net gas metering. It's a combined
+version of [EIP-1283] and [EIP-1706], with a structured definition so as
+to make it interoperable with other gas changes such as [EIP-1884].
+
+## Abstract
+
+This EIP provides a structured definition of net gas metering changes
+for `SSTORE` opcode, enabling new usages for contract storage, and
+reducing excessive gas costs where it doesn’t match how most
+implementation works.
+
+This is a combination of EIP-1283 and EIP-1706.
+
+## Motivation
+
+This EIP proposes a way for gas metering on `SSTORE`, using information
+that is more universally available to most implementations, and
+require as little change in implementation structures as possible.
+
+* Storage slot’s original value.
+* Storage slot’s current value.
+* Refund counter.
+
+Usages that benefits from this EIP’s gas reduction scheme includes:
+
+* Subsequent storage write operations within the same call frame. This
+  includes reentry locks, same-contract multi-send, etc.
+* Exchange storage information between sub call frame and parent call
+  frame, where this information does not need to be persistent outside
+  of a transaction. This includes sub-frame error codes and message
+  passing, etc.
+
+The original definition of EIP-1283 created a danger of a new kind of
+reentrancy attacks on existing contracts as Solidity by default grants
+a "stipend" of 2300 gas to simple transfer calls. This danger is
+easily mitigated if `SSTORE` is not allowed in low gasleft state,
+without breaking the backward compatibility and the original intention
+of EIP-1283.
+
+This EIP also replaces the original EIP-1283 value definitions of gas
+by parameters, so that it's more structured, and easier to define
+changes in the future.
+
+## Specification
+
+Define variables `SLOAD_GAS`, `SSTORE_SET_GAS`, `SSTORE_RESET_GAS` and
+`SSTORE_CLEARS_SCHEDULE`. The old and new values for those variables
+are:
+
+* `SLOAD_GAS`: changed from `200` to `800`.
+* `SSTORE_SET_GAS`: `20000`, not changed.
+* `SSTORE_RESET_GAS`: `5000`, not changed.
+* `SSTORE_CLEARS_SCHEDULE`: `15000`, not changed.
+
+Change the definition of EIP-1283 using those variables. The new
+specification, combining EIP-1283 and EIP-1706, will look like
+below. The terms *original value*, *current value* and *new value* are
+defined in EIP-1283.
+
+Replace `SSTORE` opcode gas cost calculation (including refunds) with
+the following logic:
+
+* If *current value* equals *new value* (this is a no-op), `SLOAD_GAS`
```

Comment:
To me it seems the geth implementation is slightly different: https://github.com/ethereum/go-ethereum/pull/19964/files#diff-d45223314f7eaec57219899f6b9297feR165

Reply:
Could you elaborate on what you think is different?

Reply:
Ah yes, we start by checking `2300`

Reply:
Moved.